docs_src/microservices/security/Ch-SSH-Tunneling-HowToSecureDeviceServices.md

```diff
@@ -0,0 +1,297 @@
+
+"Top level diagram for SSH tunneling for device services"
+
+In the primary, the SSH tunneling initiated by device-ssh-proxy service to the remote running device services.  The dependencies that remote device services needed are reversely tunneling from the primary machine.
+
+## How to- Reference implementation example
+
+The whole reference implementation example can be found in this repository: 
+[HoldingRepoSshTunneling](https://github.com/edgexfoundry-holding/external-device-security-examples/ssh-tunneling/README.md)
+
+### Setup remote running Virtual Machine
+
+In the example setup, `vagrant` is used on the top of `Virtual Box` to set up as the secondary/remote VM. The network port for ssh standard port 22 is mapped into 2222 for `vagrant ssh` itself and the forwarded port is also mapped on the VM network for port 2223 to the host machine port 2223.  This port 2223 is used for the ssh daemon Docker container that will be introduced later on below.
+
+Once you have downloaded the vagrant from Hashicorp website, typical vagrant setup for the first time can be done via command `./vagrant init` and it will generate the Vagrant configuration file.
+
+Here is the Vagrant file used to create the remote machine:
+
+[remote VM Vagrant file with docker and docker-compose installed](https://github.com/edgexfoundry-holding/external-device-security-examples/ssh-tunneling/Vagrantfile)
+
+### SSH Tunneling: Setup the SSH server on the remote machine
+
+Use the example from Docker hub, an example of ssh damenon or ssh server can be setup pretty easily as a Docker container: https://docs.docker.com/engine/examples/running_ssh_service/
+
+Note that this one is the ssh server and it is set up using password authentication by default.  In order to authenticate to this ssh server without password prompt, we injected the generated public SSH key from the primary machine via simple login into the ssh server machine first and then created the `authorized_keys` under `~/.ssh` directory.  Without loss of generality, the following command example shows how this is accomplished:
+
+```sh
+root@sshd-remote: mkdir -p ~/.ssh
+root@sshd-remote: chmod 700 ~/.ssh
+root@sshd-remote:  echo "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDKvsFf5HocBOBWXdVJKfQzkhf0K8lSLjZn9PX84VdhHyP8n1mzfpZywA4vsz8+A3OsGHAr2xpkyzOS0YkwD7nrI3q1x0A0+ANhQNOaKbnfQRe... root" >> ~/.ssh/authorized_keys
+```
+
+The ssh key pairs can be generated using `ssh-keygen` command from the primary machine and the contents of ssh public key usually is stored as ~/.ssh/id_rsa.pub file like this:
+
+```sh
+ssh-keygen -q -t rsa -C root -N '' -f ~/.ssh/id_rsa 2>/dev/null
+```
+
+One example of build script to make SSH public key be injected is given as like:
+
+[example build shell script to build and install SSH public key into sshd server](https://github.com/edgexfoundry-holding/external-device-security-examples/ssh-tunneling/build.sh)
+
+### SSH Tunneling: Local Port Forwarding
+
+from primary to secondary /remote
```

Comment:
primary -> local, remove secondary

Reply:
Impromper casing and  little unclear what is being said. Try "This is how to connect from the local machine to the remote machine. The `-L` flag is important here."